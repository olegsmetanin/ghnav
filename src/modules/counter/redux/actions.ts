export const actionTypes = {
  INCREMENT: 'COUNTER/INCREMENT',
  INCREMENT_SUCCESS: 'COUNTER/INCREMENT_SUCCESS',
  INCREMENT_FAILURE: 'COUNTER/INCREMENT_FAILURE'
}

export function increment(value: number) {
  return {
    type: actionTypes.INCREMENT,
    payload: value
  }
}

export function incrementSuccess(value: number) {
  return {
    type: actionTypes.INCREMENT_SUCCESS,
    payload: value
  }
}

export function incrementFailure(error) {
  return {
    type: actionTypes.INCREMENT_FAILURE,
    error
  }
}
