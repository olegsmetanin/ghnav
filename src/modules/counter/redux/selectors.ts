export const counterValueSelector = state => state.counter.value

export const counterErrorSelector = state => state.counter.error
