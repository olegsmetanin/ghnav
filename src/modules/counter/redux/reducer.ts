import { actionTypes } from './actions'

export const initialState = {
  value: 0
}

export function reducer(state = initialState, action) {
  switch (action.type) {
    case actionTypes.INCREMENT_SUCCESS:
      return {
        ...state,
        ...{
          value: state.value + action.payload
        }
      }

    default:
      return state
  }
}
