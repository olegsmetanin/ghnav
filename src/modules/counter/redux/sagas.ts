import { actionTypes, incrementFailure, incrementSuccess } from './actions'
import { put, takeLatest } from 'redux-saga/effects'

export function* incrementSaga(action) {
  try {
    const value = action.payload

    if (typeof value === 'number') {
      yield put(incrementSuccess(value))
    } else {
      throw new Error('increment value must be a number')
    }
  } catch (err) {
    yield put(incrementFailure(err.message))
  }
}

export const sagas = [takeLatest(actionTypes.INCREMENT, incrementSaga)]
