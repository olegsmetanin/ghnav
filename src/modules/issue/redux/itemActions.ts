import { createActions } from 'common/redux/createActions'

export const { actionTypes, failure, load, loadSuccess } = createActions(
  'ISSUE'
)
