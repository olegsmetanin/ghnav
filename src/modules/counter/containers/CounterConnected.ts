import { Counter } from '../components/Counter'
import { connect } from 'react-redux'
import { counterValueSelector } from '../redux/selectors'
import { increment } from '../redux/actions'

const mapStateToProps = state => ({
  value: counterValueSelector(state)
})

const mapDispatchToProps = dispatch => ({
  onIncrement: value => dispatch(increment(value))
})

export const CounterConnected = connect(
  mapStateToProps,
  mapDispatchToProps
)(Counter)
