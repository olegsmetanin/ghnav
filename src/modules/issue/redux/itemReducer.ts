import { actionTypes } from './itemActions'
import { handleActions } from 'common/redux/handleActions'

export const initialState = {
  error: null,
  process: {
    isLoading: true
  }
}

interface IState {
  value?: any
  error?: any
  query?: any
}

export const itemReducer = handleActions<IState>(
  {
    [actionTypes.FAILURE]: (state, action) => {
      return {
        ...state,
        ...{
          error: action.error
        }
      }
    },
    [actionTypes.LOAD_SUCCESS]: (state, action) => {
      return {
        ...state,
        ...{
          value: action.payload.value,
          query: action.payload.query
        }
      }
    }
  },
  initialState
)

// export function itemReducer(state = initialState, action) {
//   switch (action.type) {
//     case actionTypes.FAILURE:
//       return {
//         ...state,
//         ...{
//           error: action.error
//         }
//       }

//     case actionTypes.LOAD_SUCCESS:
//       return {
//         ...state,
//         ...{
//           value: action.payload.value,
//           query: action.payload.query
//         }
//       }

//     default:
//       return state
//   }
// }
