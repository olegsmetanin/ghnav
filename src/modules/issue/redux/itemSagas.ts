/* global fetch */

import 'isomorphic-unfetch'

import { actionTypes, failure, loadSuccess } from './itemActions'
import { put, takeLatest } from 'redux-saga/effects'

// import jsontab from 'modules/issue/containers/__tests__/issue.stab.json'

export function* loadSaga(action) {
  try {
    const query = action.payload
    const url = `https://api.github.com/repos/${query.owner}/${query.repo}/issues/${query.num}`
    const response = yield fetch(url)
    if (response.status < 200 || response.status > 300) {
      throw new Error(response.statusText)
    }
    const value = yield response.json()
    yield put(loadSuccess({ query, value }))
  } catch (err) {
    yield put(failure(err.message))
  }
}

export const itemSagas = [takeLatest(actionTypes.LOAD, loadSaga)]
