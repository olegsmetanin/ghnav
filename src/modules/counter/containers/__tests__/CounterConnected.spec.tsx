import { CounterConnected } from '../CounterConnected'
import React from 'react'
import { all } from 'redux-saga/effects'
import configureMockStore from 'redux-mock-store'
import { sagas as counterSagas } from 'modules/counter/redux/sagas'
import createSagaMiddleware from 'redux-saga'
/* eslint-env jest */
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'

const sagaMiddleware = createSagaMiddleware()
const mockStore = configureMockStore([sagaMiddleware])

describe('CounterConnected', () => {
  it('renders correct with click', () => {
    const store = mockStore({ counter: { value: 0 } })

    const next = () => 1

    const tree = renderer.create(
      <CounterConnected id="cc" store={store} next={next}>
        Hello
      </CounterConnected>
    )
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('failed if next returns not a number', () => {
    const store = mockStore({ counter: { value: 0 } })

    function* rootSaga() {
      yield all([...counterSagas])
    }

    store.runSagaTask = () => {
      store.sagaTask = sagaMiddleware.run(rootSaga)
    }

    store.runSagaTask()

    const next = () => 'qwe'

    const tree = mount(
      <CounterConnected id="cc" store={store} next={next}>
        Hello
      </CounterConnected>
    )

    tree
      .find('#cc')
      .first()
      .simulate('click')
    const actions = store.getActions()
    expect(actions).toEqual([
      { payload: 'qwe', type: 'COUNTER/INCREMENT' },
      { error: 'increment value must be a number', type: 'COUNTER/INCREMENT_FAILURE' }
    ])
  })
})
