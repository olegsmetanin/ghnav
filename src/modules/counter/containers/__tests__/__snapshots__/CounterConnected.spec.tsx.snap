// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterConnected renders correct with click 1`] = `
<span
  onClick={[Function]}
>
  Hello
   
  0
</span>
`;
